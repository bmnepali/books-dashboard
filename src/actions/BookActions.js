import { BOOKS_LOAD, BOOKS_LOADED, BOOKS_ERROR, CHANGE_CATEGORY } from '../constants/BookActionTypes';
import BooksService from '../services/Book.service';
import store from '../store';

// Book Actions
export function loadingBooks() {
  return { type: BOOKS_LOAD };
}

export function booksLoaded(payload) {
  return {
    type: BOOKS_LOADED,
    payload: payload,
  };
}

export function errorLoadingBooks(error) {
  return {
    type: BOOKS_ERROR,
    payload: error,
  };
}

export function changeCategory(category) {
  return {
    type: CHANGE_CATEGORY,
    payload: category,
  };
}

export function loadBooks(category) {
  store.dispatch(loadingBooks());
  BooksService.getBooks(category)
    .then(response => {
      store.dispatch(booksLoaded(response.data.results));
    })
    .catch(error => {
      errorLoadingBooks(error);
    });
}
