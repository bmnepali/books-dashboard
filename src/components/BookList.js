import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Book from './Book';

class BookList extends Component {
  constructor(props) {
    super(props);
  }

  // Executes after the component is mounted
  componentDidMount() {
    this.props.loadBooksFromApi(this.props.booksStore.category);
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.booksStore.category !== nextProps.booksStore.category) {
      this.props.loadBooksFromApi(nextProps.booksStore.category);
    }
  }

  render() {
    let view = null;

    if (this.props.booksStore.isLoading) {
      view = (
        <div className="loader">
          <div className="spinner" />
        </div>
      );
    } else if (this.props.booksStore.error) {
      view = (
        <div className="alert alert-danger" role="alert">
          <strong>Hey there!</strong> {this.props.booksStore.error}
        </div>
      );
    } else {
      if (!this.props.booksStore.isLoading && this.props.booksStore.books.length) {
        view = (
          <ul className="row book-list">
            {this.props.booksStore.books.map(book => (
              <li key={book.amazon_product_url} className="col-lg-6 item-book">
                <Book book={book} />
              </li>
            ))}
          </ul>
        );
      } else {
        view = (
          <div className="no-books-container">
            <i className="icon ion-md-sad" />
            No Books found!
          </div>
        );
      }
    }

    return view;
  }
}

BookList.propTypes = {
  loadBooksFromApi: PropTypes.func.isRequired,
  booksStore: PropTypes.shape({
    isLoading: PropTypes.bool.isRequired,
    books: PropTypes.arrayOf(PropTypes.object).isRequired,
    error: PropTypes.string,
    category: PropTypes.string.isRequired,
  }).isRequired,
};

export default BookList;
