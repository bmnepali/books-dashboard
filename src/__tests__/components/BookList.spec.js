import * as React from 'react';
import axios from '../../utilities/api';
import BookService from '../../services/Book.service';
import BookList from '../../components/BookList';
import { loadBooks } from '../../actions/BookActions';

describe('BookList Component', () => {
  const props = {
    loadBooksFromApi: loadBooks,
    booksStore: {
      isLoading: false,
      books: [],
      error: null,
      category: 'fiction',
    }
  };

  describe('when rendering the component', () => {
    it('renders the component without any error', () => {
      shallow(<BookList {...props} />);
    });

    it('matches the snapshot', () => {
      expect(toJson(shallow(<BookList {...props} />))).toMatchSnapshot();
    });
  });

  describe('when there are no books', () => {
    it('displays no books found message as there are no books in the list initially', () => {
      const component = shallow(<BookList {...props} />);
      expect(component.find('.no-books-container').length).toBe(1);
      expect(component.find('.no-books-container').text()).toBe('No Books found!');
    });
  });

  describe('when there are books', () => {
    it('displays books list if there are books in the list', () => {
      props.booksStore.books = [{
        amazon_product_url: 'https://some-path-one',
        book_details: [{
          title: 'Book One',
          description: 'Book One Description',
          primary_isbn13: 759723497592
        }]
      },
      {
        amazon_product_url: 'https://some-path-two',
        book_details: [{
          title: 'Book Two',
          description: 'Book Two Description',
          primary_isbn13: 231423480984
        }]
      }];

      const component = shallow(<BookList {...props} />);
      expect(component.find('li').length).toBe(2);
    });
  });

  describe('when the component is mounted ', () => {
    it('calls BookService`s getBooks() with expected category as parameter', () => {
      const spy = jest.spyOn(BookService, 'getBooks');
      const wrapper = shallow(<BookList {...props} />);

      wrapper.instance().forceUpdate();
      expect(spy).toHaveBeenCalledWith(props.booksStore.category);
    });

    it('calls axios`s get() with expected books api path', () => {
      const spyAxios = jest.spyOn(axios, 'get');
      const wrapper = shallow(<BookList {...props} />);

      wrapper.instance().forceUpdate();
      expect(spyAxios).toHaveBeenCalledWith('lists.json?list-name=fiction&api-key=6ad84e249d054efeaefe1abb8f89df5b');
    });
  });
});

