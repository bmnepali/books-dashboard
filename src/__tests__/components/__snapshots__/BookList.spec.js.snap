// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList Component when rendering the component matches the snapshot 1`] = `
<div
  className="no-books-container"
>
  <i
    className="icon ion-md-sad"
  />
  No Books found!
</div>
`;
